src/QsCompiler/SyntaxProcessor/TypeInference/Constraint.fs

```diff
@@ -0,0 +1,37 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.QsCompiler.SyntaxProcessing.TypeInference
+
+open Microsoft.Quantum.QsCompiler.SyntaxTokens
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+
+type internal Constraint =
```

Comment:
I am a curious about why the `Constraint` type is defined in two separate source files?

Reply:
Constraint.fsi is the signature file for Constraint.fs. In general, the signature file contains only the types of things in the implementation file (so Constraint.fsi has `val types: Constraint -> ResolvedType list`, while Constraint.fs has `let types = ...`).

For discriminated unions and records with public or internal cases/fields, the signature and implementation are actually the same, but it's still required to provide them separately, so unfortunately they are duplicated across both files.